src/main/java/org/spongepowered/api/util/raytrace/RayTracingUtil.java

```diff
@@ -0,0 +1,160 @@
+     *
+     * @param world The world to search in
+     * @param consumer The consumer that consumes all visited blocks
+     * @return The wrapped consumer
+     */
+    public static PositionConsumer wrapSolidBox(Extent world, BlockConsumer consumer) {
+        return SOLIDBOX_BLOCKADAPTER.wrap(world, consumer);
+    }
+
+    /**
+     * Wraps the given BlockConsumer in a {@link PositionConsumer}. This tries
+     * to check whether the interact box of the block has been hit. This wrapper
+     * makes some assumptions on blocks, so it might not act the same as the
+     * client.
+     *
+     * @param world The world to search in
+     * @param consumer The consumer that consumes all visited blocks
+     * @return The wrapped consumer
+     */
+    public static PositionConsumer wrapInteractBox(Extent world, BlockConsumer consumer) {
+        return INTERACTBOX_BLOCKADAPTER.wrap(world, consumer);
+    }
+
+    /**
+     * Starts the ray tracing from the given location.
+     *
+     * <p>Note: The range may be truncated by server settings.</p>
+     *
+     * @param location The location to start
+     * @param direction The ray direction
+     * @param consumer The consumer that consumes all visited blocks
+     */
+    public static void trace(Vector3d location, Vector3d direction, PositionConsumer consumer) {
+        double posX = location.getX();
+        double posY = location.getY();
+        double posZ = location.getZ();
+        // Does not move?
+        if (direction.lengthSquared() == 0) {
+            consumer.apply(posX, posY, posZ, direction);
+            return;
+        }
+        final Vector3d normalizedDirection = direction.normalize();
+        final double dx = normalizedDirection.getX();
+        final double dy = normalizedDirection.getY();
+        final double dz = normalizedDirection.getZ();
+        final double[] distances = new double[3];
+        final PositionConsumer limitedConsumer = POSITION_LIMITER.apply(consumer);
+        while (limitedConsumer.apply(posX, posY, posZ, direction)) {
+            distances[0] = distance(posX, dx);
+            distances[1] = distance(posY, dy);
+            distances[2] = distance(posZ, dz);
+            final double min = min(distances);
+            posX = sanitize(posX + min * dx);
+            posY = sanitize(posY + min * dy);
+            posZ = sanitize(posZ + min * dz);
+        }
+    }
+
+    // The minimum distance
+    private static double min(double[] moves) {
+        double min = moves[0];
+        if (min > moves[1]) {
+            min = moves[1];
+        }
+        if (min > moves[2]) {
+            min = moves[2];
+        }
+        return min;
+    }
+
+    // times how often you have to apply the given direction to pass any borders
+    private static double distance(double position, double direction) {
+        if (direction == 0) {
+            return Double.MAX_VALUE;
+        }
+        final double distanceLeft = (position % 1 + 1) % 1;
```

Comment:
Because this method deals with directions, distanceLeft should be renamed or commented to diversify it from the direction left. Perhaps rename it to "distanceRemaining"